### PR TITLE
Add notes on tangles

### PR DESCRIPTION
**Pull request by mixmix**

### CODE REVIEW

**Comment by Powersource** [comment]
Was gonna say "we can close this now" but that other PR only contains general tangle stuff. The stuff about how the group tangles work should be put somewhere else I guess.

File: README.md [hunk 1]
```diff
@@ -498,17 +498,341 @@ group members to eventually converge on a single most preferred epoch.  We have
...
+        - check that the graph contains each message in `previous`
+            - if it does, add it to the graph
+            - otherwise set it aside till the graph does contain them
+        - newly added messages to the tangle graph constitue new tips
+    - c) repeat (b) till you cannot grow the tangle further
+
+
+In the context of private groups, the addition to a particular epoch is always
+published to an earlier epoch (or in the case of the first epoch, to the
+Additons feed).
```
Comment (Powersource) [inline README.md:629]: Only the re-additions. Newly added members will still get their invites on the additions feed. Oh yeah that's another case we have to test. Is that even in the spec? I.e. getting access to old epochs if getting added to a later one.

Made an issue https://github.com/ssbc/ssb-tribes2/issues/76

**Comment by staltz** [comment]
> The stuff about how the group tangles work should be put somewhere else I guess.

It should be put in this PR or this repo, IMO.

File: README.md (hunk 1 above)
Comment (mixmix) [reply README.md:536]: Disagree. Newly added people ... if there are messages `A -> B -> C` should receive invites to :
- `A` - invite published in the Additions feed
- `B` - invite published in `A`
- `C` - invite published in `B`

we do this so that we can easily query the membership tangle in each epoch and not have some funny "special case" for people added to epochs later

**Comment by mixmix** [comment]
@staltz I removed the tangle stuff that is now in the SIP#9 (tangle sip)
I think this is ready to merge now.

Can you please review,and merge if ready?

File: README.md [hunk 2]
```diff
@@ -498,17 +498,248 @@ group members to eventually converge on a single most preferred epoch.  We have
...
+We define the members tangle for some id `A` as:
+1. **candate messages**:
+   - the root message with id `A` so long as it has
+     ```javascript
+     content.tangles.members = { root: null, previous: null }
+     ```
+   - any messages where
+     ```javascript
+     content.tangles.members = { root: A, previous: [...] }
+     ```
```
Comment (staltz) [inline README.md:521]: If I understood tangles correctly, the "candidates" are the input, and the "recipe" generates the output, which are the fields `content.tangles.members`. So it's not correct to put `content.tangles.members` as the candidates, because those are precisely the outcome of the recipe!

File: README.md [hunk 3]
```diff
@@ -498,17 +498,248 @@ group members to eventually converge on a single most preferred epoch.  We have
...
+   - the root message with id `A` so long as it has
+     ```javascript
+     content.type = 'group/init'
+     content.tangles.epoch = { root: null, previous: null }
+     ```
+   - any messages where
+     ```javascript
+     content.type = 'group/init'
+     content.tangles.epoch = { root: A, previous: [...] }
+     ```
```
Comment (staltz) [inline README.md:575]: The above has the same "recipes output as candidates" problem

File: README.md [hunk 4]
```diff
@@ -498,17 +498,248 @@ group members to eventually converge on a single most preferred epoch.  We have
...
+1. **candate messages**:
+   - the root message with id `A` so long as it has
+     ```javascript
+     content.type = 'group/init'
+     content.tangles.group = { root: null, previous: null }
+     ```
+   - any messages where
+     ```javascript
+     content.tangles.group = { root: A, previous: [...] }
+     ```
```
Comment (staltz) [inline README.md:623]: The above has the same "recipes output as candidates" problem

File: README.md [hunk 5]
```diff
@@ -498,17 +498,248 @@ group members to eventually converge on a single most preferred epoch.  We have
...
+
+classDef default stroke:none,fill:#ccf,color:#333;
+classDef cluster fill:#fff,stroke:#000,color:#333;
+````
+-->
 
-### 6.1. Normative References
+
+### 6.4. Using all the tangles together
+
```
Comment (staltz) [inline README.md:650]:
```suggestion

In this section we provide an example that illustrates the presence of all three group-related tangles.
Suppose peer Z creates a group and performs the following actions, in this order:

1. Adds peer A to the group at epoch 0
2. Adds peers B and C to the group at epoch 0
3. Excludes peer C, thus creating epoch 1
4. Re-adds peers A and B to the group, now at epoch 1

Then the following diagram illustrates Z's feeds (additions feed, epoch 0 feed, epoch 1 feed) and its
messages:
```

**Review by staltz — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (staltz) [reply README.md:536]: This is important, and I think @Powersource is right. If I invite my friend to epoch N, I should NOT publish the invite on epoch N-1, because my "group feed" for epoch N-1 is unknown/invisible to anyone who is outside of the group. My friend won't even replicate that group feed, so how could they even see my invite?

Thus it has to be on the additions feed, because this one is publicly known.

**Comment by staltz** [comment]
@mixmix I reviewed, and found important things to address before merging.

File: README.md (hunk 1 above)
Comment (Powersource) [reply README.md:536]: @mixmix are you saying, when adding a new person, we should give them one invite per epoch? that feels a bit unnecessary when we could just invite them to the latest and point them backwards. i'd say when we're on a new epoch, the old member tangle on the old epoch is no longer representative for the current state of the group. if it was representative, then we'd have to keep updating old epochs forever, which would also be annoying for replication

Comment (Powersource) [reply README.md:536]: would like this discussion to be resolved before merging, or at least marked as a :fire: TODO for later

File: README.md (hunk 5 above)
Comment (mixmix) [reply README.md:650]: why do you want to remove this helpful description of what the diagram shows?

Comment (staltz) [reply README.md:650]: Remove?? My commit suggestion is *adding* the description.

Comment (mixmix) [reply README.md:650]: weird, my brain was poo apparently

File: README.md (hunk 2 above)
Comment (mixmix) [reply README.md:521]: I don't understand. The tangles data in `content.tangles.members` is the data needed for building the tangle (following the recipe). And if you want to add a new messages, then yes you bake a new message with something which will be consistent with the recipe outcome sofar?

File: README.md [hunk 6]
```diff
@@ -498,17 +498,248 @@ group members to eventually converge on a single most preferred epoch.  We have
...
+A "tangle" in scuttlebutt is a way to define a directed acyclic graph (DAG) of 
+messages. You can read more about them in the [Tangle SIP].
+
+
+### 6.1. Members tangle
+
+The purpose of this tangle is to track the group membership (for a  particular 
+epoch of the group).
+
+We define the members tangle for some id `A` as:
```
[the remark this answers is not in the document]
Comment (mixmix) [reply README.md:512]: I skipped definining what A was because it's specified in Candidate messages

File: README.md [hunk 7]
```diff
@@ -498,17 +498,248 @@ group members to eventually converge on a single most preferred epoch.  We have
 not found a tie-breaking rule with all three properties, so this is future work.
 
 
-## 6. References
+## 6. Tangles
```
[the remark this answers is not in the document]
Comment (mixmix) [reply README.md:501]: ok, moving now

File: README.md [hunk 8]
```diff
@@ -468,6 +468,248 @@ From this we can see `e` must be added to epoch `Z` to bring it up to "correct
...
+A "tangle" in scuttlebutt is a way to define a directed acyclic graph (DAG) of 
+messages. You can read more about them in the [Tangle SIP].
+
+
+### 4.10.1. Members tangle
+
+The purpose of this tangle is to track the group membership (for a  particular 
+epoch of the group).
+
+We define the members tangle for some id `A` as:
```
Comment (mixmix) [inline README.md:482]:
```suggestion
We define the members tangle for some message ID `A`as:
```

File: README.md (hunk 2 above)
Comment (mixmix) [reply README.md:491]: the recipe generates a tangle-graph. You can read the tips of this to generate the `previous` for any "next message"

Comment (mixmix) [reply README.md:491]: @staltz can we chat about this sync - text clarification of this feels too slow

Comment (staltz) [reply README.md:491]: Yeah, the way you think of candidates is very confusing to me, and I bet I won't be the only one. Let's talk sync.

**Comment by staltz** [comment]
High-level notes on what this PR needs before I can approve it:

- "Candidates chicken-and-egg" problem
- Sprinkling MUSTs and SHOULDs (this is important, what parts of the spec are nice-to-have and which parts are non-negotiable? That's what MUST and SHOULD informs the reader)

File: README.md [hunk 9]
```diff
@@ -470,40 +470,36 @@ membership".
...
+the Set of members in an epoch by [reducing] the messages in this tangle.  Each
+epoch will have its own members tangle.  The human-friendly name in the tangle
+data is the string `members`.
+
+We construct the members tangle for some epoch `E` of a group `G` as:
+
+1. **Candidate messages**:
+    - The message with `content.type` equal to `group/init` which started epoch
+    `E`
+    - Any message published on the epoch feed for `E` where `content.type` is
```
Comment (mixmix) [inline README.md:490]: "epoch feed" is vague. I think your thinking about pppp is leaking in here!

If "feed" is redefined somewhere else then it's fine, but otherwise please revert to mentioning

```js
content.tangles.members = { root: E.id , previous: [...] }
```

note you might need to define the what `E.id` means....

File: README.md [hunk 10]
```diff
@@ -470,40 +470,36 @@ membership".
...
+The purpose of the members tangle is to group and partially order all messages
+that alter the group membership during a single epoch.  A peer can calculate
+the Set of members in an epoch by [reducing] the messages in this tangle.  Each
+epoch will have its own members tangle.  The human-friendly name in the tangle
+data is the string `members`.
+
+We construct the members tangle for some epoch `E` of a group `G` as:
+
+1. **Candidate messages**:
+    - The message with `content.type` equal to `group/init` which started epoch
```
Comment (mixmix) [inline README.md:488]: epoch `E` and this root message are related directly by `E.key` or `E.id` ... I feel like this removes clarity

File: README.md [hunk 11]
```diff
@@ -470,40 +470,36 @@ membership".
...
+    - Any message published on the epoch feed for `E` where `content.type` is
+    equal to `group/add-member`
+    - Any message published on the epoch feed for `E` where `content.type` is
+    equal to `group/exclude`
+2. **Recipe**:
+    - If the candidate message has `content.type` equal to `group/init`, then
+    the tangle data MUST be `{"root": null, "previous": null}`.
+    - Otherwise, the candidate message's tangle data MUST have `root` equal to
+    the ID of the root message `R`, and `previous` SHOULD be an array of message
+    IDs for all locally known "tips" of the tangle
```
Comment (mixmix) [inline README.md:499]: To me, these are candidacy requirements. The "recipe" is how you take candidates and build a tangle from them.
- do you just follow the "basic" recipe?
- are there any additional checks you run while extending the graph

The candidacy for me is about more than the type (in the code). I drop any malformed messages - e.g. if you don't have a recps field on `group/add-member` then that will not even make it to the recipe step.

**Review by mixmix — COMMENTED** [review]
REQUEST CHANGES (it won't let me select that)

I thought we were going to split out:
- how to read a tangle (built it)
- how to write a new message into a tangle

Dunno if that was for this repo or another but this one has conversation about chicken/egg, so yeah. As I said, the edits feel a bit unclear to me. Exhausted after wake-ups from kids last night (I even got to bed early, only to be punished by tiny demons). Happy to discuss and push this to done together.

**Comment by staltz** [comment]
> I thought we were going to split out:
> - how to read a tangle (built it)
> - how to write a new message into a tangle
> 
> Dunno if that was for this repo or another but this one has conversation about chicken/egg, so yeah.

Yes I was thinking about the split. It - [ ] occurred to me that how to write a tangle is more prescriptive (more speccy) than how to read one. Arguably there are multiple valid ways of reading a tangle: you could just group all tangle messages, and *count* them (e.g. if you're trying to discover how many epochs have existed), without topologically sorting them, or then you may construct a graph data structure and topo sort and do other things.

In the introductory paragraphs I hinted at how the tangle is useful when read. But the candidates and recipes concerns the write part.

If we're stuck, we can either ask a 3rd party perspective to tell us if this is clear or not, and what needs clarity. Or, we do another video call.

**Comment by mixmix** [comment]
I see reading and writing as really related @staltz - knowing what "recipe" you're following (candidates + how to combine?) is crucial to you and I taking the same N messages and being able to "read" the state and get a deterministic answer. If you don't bother with a topo graph, then you are essentially having a different conversation than me. We see this with e.g. Planetary + Patchfox just making `branch` whatever message they feel like. They didn't read the state the same, and so when they speak they speak .... out of order / in a confusing way, so that when others try to read, their messages are all in the wrong place.

**To summarize**: yes you can read a tangle however you like, but for meaningful interaction there must be a canonical pattern/recipe for each tangle type.

I cannot see what it is that you and I don't share. I feel like there is something we're missing alignment on. Because I keep sharing things which I think are clear and you say "yes", but then say things which wouldn't make sense if I had successfully conveyed my thought?

**Next step:** yeah I don't know. Another person / Another call both feel costly at the moment. Can we just do an ~OK spec which neither of us is happy with but gets all the ideas on the page?

**Review by staltz — APPROVED** [review]

**Comment by staltz** [comment]
@mixmix Ready for another review! https://github.com/ssbc/ssb-group-exclusion-spec/blob/tangles/README.md

File: README.md [hunk 12]
```diff
@@ -484,19 +484,35 @@ data is the string `members`.
...
+        `group/exclude` published on an epoch feed for `E`, (2) have `ROOTID`
+        in the `root` field of their tangle data, (3) contain this tip in the
+        `previous` field of their tangle data, (4) all messages in the
+        `previous` field are in the "tangle nodes"
+        - If there are any such messages, then:
+          - Remove the tip from the "tip set"
+          - Add each such message to the "tip set"
+          - Add each such message to the "tangle nodes"
+    - d) Repeat (c) till you cannot update the tip set further
+    - e) The remaining messages in the "tip set" are the tips of the tangle
```
Comment (mixmix) [inline README.md:514]: This is clear ! :heart:

**Review by mixmix — COMMENTED** [review]
Hearty Approval!

**Comment by staltz** [comment]
Let's go!
